flight/PiOS/Common/pios_crossfire.c

```diff
@@ -0,0 +1,335 @@
+		uint16_t *headroom, bool *task_woken);
+/**
+ * @brief Reset the internal receive buffer state
+ * @param[in] dev device driver instance pointer
+ */
+static void PIOS_Crossfire_ResetBuffer(struct pios_crossfire_dev *dev);
+/**
+ * @brief Unpack a frame from the internal receive buffer to the channel buffer
+ * @param[in] dev device driver instance pointer
+ */
+static void PIOS_Crossfire_UnpackFrame(struct pios_crossfire_dev *dev);
+/**
+ * @brief RTC tick callback
+ * @param[in] context Driver instance handle
+ */
+static void PIOS_Crossfire_Supervisor(uintptr_t context);
+
+// public
+const struct pios_rcvr_driver pios_crossfire_rcvr_driver = {
+	.read = PIOS_Crossfire_Read,
+};
+
+
+static struct pios_crossfire_dev *PIOS_Crossfire_Alloc(void)
+{
+	struct pios_crossfire_dev *dev = PIOS_malloc_no_dma(sizeof(struct pios_crossfire_dev));
+	if (!dev)
+		return NULL;
+
+	memset(dev, 0, sizeof(*dev));
+	dev->magic = PIOS_CROSSFIRE_MAGIC;
+
+	return dev;
+}
+
+static bool PIOS_Crossfire_Validate(const struct pios_crossfire_dev *dev)
+{
+	return dev && dev->magic == PIOS_CROSSFIRE_MAGIC;
+}
+
+int PIOS_Crossfire_Init(uintptr_t *crsf_id, const struct pios_com_driver *driver,
+		uintptr_t uart_id)
+{
+	struct pios_crossfire_dev *dev = PIOS_Crossfire_Alloc();
+
+	if (!dev)
+		return -1;
+
+	*crsf_id = (uintptr_t)dev;
+
+	PIOS_Crossfire_ResetBuffer(dev);
+	PIOS_Crossfire_SetAllChannels(dev, PIOS_RCVR_INVALID);
+
+	if (!PIOS_RTC_RegisterTickCallback(PIOS_Crossfire_Supervisor, *crsf_id))
+		PIOS_Assert(0);
+
+	(driver->bind_rx_cb)(uart_id, PIOS_Crossfire_Receive, *crsf_id);
+
+	return 0;
+}
+
+static int32_t PIOS_Crossfire_Read(uintptr_t context, uint8_t channel)
+{
+	if (channel > PIOS_CROSSFIRE_CHANNELS)
+		return PIOS_RCVR_INVALID;
+
+	struct pios_crossfire_dev *dev = (struct pios_crossfire_dev *)context;
+	if (!PIOS_Crossfire_Validate(dev))
+		return PIOS_RCVR_NODRIVER;
+
+	return dev->channel_data[channel];
+}
+
+static void PIOS_Crossfire_SetAllChannels(struct pios_crossfire_dev *dev, uint16_t value)
+{
+	for (int i = 0; i < PIOS_CROSSFIRE_CHANNELS; i++)
+		dev->channel_data[i] = value;
+}
+
+static uint16_t PIOS_Crossfire_Receive(uintptr_t context, uint8_t *buf, uint16_t buf_len,
+		uint16_t *headroom, bool *task_woken)
+{
+	struct pios_crossfire_dev *dev = (struct pios_crossfire_dev *)context;
+
+	if (!PIOS_Crossfire_Validate(dev))
+		goto out_fail;
+
+	for (int i = 0; i < buf_len; i++) {
+		// Ignore any incoming stuff until we're expecting new data.
+		if(dev->buf_pos >= dev->bytes_expected)
+			continue;
+
+		dev->u.rx_buf[dev->buf_pos++] = buf[i];
+
+		if(dev->buf_pos == CRSF_LEN_IDX) {
+			// Read length field and adjust. Denotes payload, plus type field, plus CRC field.
+			dev->bytes_expected = CRSF_ADDRESS_LEN + CRSF_LENGTH_LEN + dev->u.frame.length;
+
+			// If length field isn't plausible, ignore rest of the data.
+			if(dev->bytes_expected >= CRSF_MAX_FRAMELEN)
+				dev->bytes_expected = 0;
```

Comment:
bytes_expected gets reset to CRSF_MAX_FRAMELEN via PIOS_Crossfire_ResetBuffer() in the supervisor during the inter-frame communication pause. Typical distance between frame starts is 6.67ms, sending a maximum frame takes around 0.8ms (RC channels is 0.56ms), which leaves plenty of time for the supervisor to do that up to 3 times. This should work well enough, right?